### PR TITLE
Form Refactor Part 2

### LINKED ISSUE #820: Forms (in general) - enhancements
- [ ] I think for all input fields we can add the option to have Helper text, which can be enabled or disabled.
image

- [ ] Add width controls & align options for the Submit button (for example if I want to make the form button full-width). Making sure the buttons inherits the styles either from Customizer or other FSE specific settings.
- [ ] Options to Control the text color for the whole form, and maybe the button style.
Inputs field are relatively ok, but imagine using the current form on a Dark Background. How as a user do I make the Labels visible?
Options to Control the text color for the whole form, and maybe the button style.
Inputs field are relatively ok, but imagine using the current form on a Dark Background. How as a user do I make the Labels visible?

- [ ] Change the required word to * symbol, instead of required word. The asterisk is a more common symbol, and it takes less time to process the informations.
Initially the asterisk could have the same color as the text, but in future we can add the option to control that.
image
___

- [ ] At some point, maybe add the option to modify the Succes & Error messages in Forms.

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
Closes #820.

### Summary

Design tweaks
Add cc & bcc

### Screenshots 

image

----

### Test instructions

1. Insert a form block.
2. Customize it
3. Set the settings in Form Options
4. Test the email sending

---- 

### Checklist before the final review

- [ ] Visual elements are not affected by independent changes.
- [ ] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [ ] It loads additional script in frontend only if it is required.
- [ ] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [ ] In case of deprecation, old blocks are safely migrated.
- [ ] It is usable in Widgets and FSE.

### CODE REVIEW

**Comment by Soare-Robert-Daniel** [comment]
@JohnPixle you can try this version so if it's right. The CSS file is [here](https://github.com/Codeinwp/otter-blocks/blob/form/small-improvments/src/blocks/blocks/form/style.scss) to check the defaults value.

File: src/blocks/blocks/form/input/block.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
 	"category": "themeisle-blocks",
 	"description": "Display a contact form for your clients.",
 	"keywords": [ "input", "text", "email" ],
-	"parent": [ "themeisle-blocks/form" ],
+	"ancestor": [ "themeisle-blocks/form" ],
```
Comment (Soare-Robert-Daniel) [inline src/blocks/blocks/form/input/block.json:9]: @HardeepAsrani @JohnPixle @mghenciu  by using this, we can allow the other blocks (section/columns/ and everything with an inner block) to insert the inputs.

With this, we can do the multiple inputs on a row.

image

The problem: this is a WP 6.0 feature. On the lower version, this field is ignored, and the inputs can be inserted from any block, not just those that are subordinated to the Form block.

Do you think this is a good compromise?

Comment (HardeepAsrani) [reply src/blocks/blocks/form/input/block.json:9]: We can just enable it for users on 6.0 and use the parent for below 5.9, I think that compromise is good enough for me. What do you think @JohnPixle?

I'm more concerned if you think this is the right way of having form designs by using Columns and other blocks inside. What do you think?

**Comment by JohnPixle** [comment]
Thanks @Soare-Robert-Daniel 
Things look good, just a couple of notes:

- [ ] **Text Size Controls**:  The values are not 100% right. They should be `14 / 16 / 18 / 20` instead of `14 / 16 / 16 / 20`. Feel free to refer to [the table](https://github.com/Codeinwp/otter-blocks/issues/820#issuecomment-1164322965) from the other issue. Also, possible to have these in **EMs** instead of **PX**?

Screenshot 2022-06-24 at 11 08 22


- [ ] Possible to have the **labels in normal weight** instead of bold by default? The user can make the bold from the toolbox popup.
- [ ] Please add a **default padding for input fields of 8px.**

Let me know,  thanks for the update!

File: src/blocks/blocks/form/input/block.json (hunk 1 above)
Comment (HardeepAsrani) [reply src/blocks/blocks/form/input/block.json:9]: I think for this release we can just allow it to use other blocks inside the form if a user is using 6.0 or above and leave it at that as you @mghenciu  and tackle the 50% issue is some other release. This way we can get this done for a release + make it possible to have that layout using the core columns.

In the future, it just might make sense to go with this as having a more complex layout system will be a pain in the CSS as we can't afford to update the HTML many times so using an outside block might be an easier way but we can figure that out some other time. 😄

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Just to be sure, it has the option to use a custom value with units, right? Like this:
Screenshot 2022-06-24 at 2 54 03 PM

**Comment by Soare-Robert-Daniel** [comment]
> @Soare-Robert-Daniel Just to be sure, it has the option to use a custom value with units, right? Like this: Screenshot 2022-06-24 at 2 54 03 PM

No, the component is just a fancy dropdown.[ Source code](https://sourcegraph.com/github.com/WordPress/gutenberg@HEAD/-/blob/packages/components/src/font-size-picker/stories/index.js?L9:8)

**Comment by Soare-Robert-Daniel** [comment]
It has the same form as the one from the Paragraph but is not accessible via props.

image

**Comment by Soare-Robert-Daniel** [comment]
> Something [like this](https://wordpress.github.io/gutenberg/?path=/story/components-fontsizepicker--with-complex-css-values) for example? or [this one](https://wordpress.github.io/gutenberg/?path=/story/components-fontsizepicker--with-units)?

The problem is that you don't have access to the slider to choose a custom value. This is the pain point.

File: src/blocks/blocks/form/input/block.json (hunk 1 above)
Comment (JohnPixle) [reply src/blocks/blocks/form/input/block.json:9]: I also do not believe that using this method (sections) is the best way to handle the 50% issue, but on the other hand I understand the compromises we need to make for Gutenberg. 

As a user, there is no question that I would expect a more industry-standard way to manage field widths. This method feels like a hack :P

I understand Hardeep's concerns however. 
Still thinking..

**Comment by Soare-Robert-Daniel** [comment]
@JohnPixle, my bad. We can add the custom units. Made a push to use EM instead PX.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Make sure to update the CSS generation and all other places where it's being used in the form.

**Comment by Soare-Robert-Daniel** [comment]
> @Soare-Robert-Daniel Make sure to update the CSS generation and all other places where it's being used in the form.

Are you talking about the CSS change in this #748 ? This will be done when that PR will be merged.

**Comment by JohnPixle** [comment]
@Soare-Robert-Daniel Thanks for the update and for taking care of the EMs :)
I only noticed the following:

- [ ]  **The Text size control for labels does not have any effect** on them, regardless which size I choose (even if I choose a custom value)

You think you can take a look and see if we can fix this? Thanks in advance!

**Comment by JohnPixle** [comment]
☝🏻 Just a heads-up that I just updated the comment, since I also noticed that the button text was having the same issue

**Comment by JohnPixle** [comment]
@Soare-Robert-Daniel I think we are good for now, what do you think? Thanks for your patience on the extra revisions, looking forward to use the form on plenty of cases in starter sites.
If @mghenciu would like to do a quick final review at some point on the UI of the form and share his input on the default styles, it would be great.
Thanks

**Comment by mghenciu** [comment]
Overall, the new look for the Form Sidebar, looks really solid, @JohnPixle & @Soare-Robert-Daniel  🚀 
And will probably get even better with the Sidebar redesign, in future.

___
I've tested with the Build from this PR and there are some issues, but I wanted **to clarify, is this just about the Sidebar UI? or the whole form block?** so I know where to focus:

- [ ] the text area **Default State**, doesn't look right:
Screenshot 2022-06-27 at 15 23 05

- [ ] Backend vs Frontend
Screenshot 2022-06-27 at 15 26 04

___

- [ ] Color Stacking in the Sidebar color tab, goes under the chevron icon:
Screenshot 2022-06-27 at 15 41 10

**Comment by mghenciu** [comment]
> This might be related to the browser, something didn't load correctly (possibly the CSS file).

Created a clean instance now, and same thing happens both in Chrome and Safari:
image

 Admin area URL: https://needyhobbies.s3-tastewp.com/wp-admin 
 Username: admin 
 Password: a4FiVa3fTVM 
 
 
 ___
For the Colors, can't we disable the Stacking for this case? to avoid the overflow

**Comment by Soare-Robert-Daniel** [comment]
> Created a clean instance now, and the same thing happens both in Chrome and Safari:

Checked your instance, and the CSS files are not loaded. It might be related to TesteWP. I will update the branch and see if its happens again.

> For the Colors, can't we disable the Stacking for this case? to avoid the overflow

It seems that it does not have the option.

**Comment by JohnPixle** [comment]
Hey! I am away from screen mostly today but i also want to comment that the form did not look that way in my instance, definitely an issue with CSS or something else.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Could be an issue with the build file. You can check and regenerate it if needed.

**Comment by Soare-Robert-Daniel** [comment]
It seams that there no CSS files with the code for blocks in build. @HardeepAsrani can you take a look?

image

**Comment by Soare-Robert-Daniel** [comment]
@JohnPixle @mghenciu this a zip for testing the Form with the split color

[otter-blocks-dev.zip](https://github.com/Codeinwp/otter-blocks/files/8998634/otter-blocks-dev.zip)

@HardeepAsrani `npm run build-dev` does not put the CSS for the blocks. I made the ZIP with the output of `npm run start`

**Comment by mghenciu** [comment]
> @JohnPixle @mghenciu this a zip for testing the Form with the split color
> 
> [otter-blocks-dev.zip](https://github.com/Codeinwp/otter-blocks/files/8998634/otter-blocks-dev.zip)
> 

So should this build have different Tabs for Colors? @Soare-Robert-Daniel 
Not sure if I am looking in the right place: 
Screenshot 2022-06-28 at 10 41 57

**To mitigate this, for now** (until we deploy the Sidebar redesign), I think we can split the colors in 2: Button Colors (with the button settings) & Form Colors (the remaining color settings).

**Comment by Soare-Robert-Daniel** [comment]
Uploaded the wrong zip 
[otter-blocks-dev.zip](https://github.com/Codeinwp/otter-blocks/files/8998813/otter-blocks-dev.zip)

**Comment by mghenciu** [comment]
The UI Looks good, @Soare-Robert-Daniel !
My only suggestion would be to change the Button Style keyword to **Alignment**, as the style keyword is usually used for something else.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested and everything seems to work fine, only some small things here: 

- [ ] The name selected in **From Name** [option](https://vertis.d.pr/SbayeQ) will have the last letter truncated in the sent email. If i set it as `Irinel`, it will appear as `Irine`
- [ ] When testing on iphone, the spacings are not applied anymore and the date field is smaller until i set some value into it
How it looks on iphone: https://vertis.d.pr/WNU3dX
How it looks on android: https://vertis.d.pr/AqguPI
(I could replicate it only on Browserstack using iphone 11, i don't know if it's a real problem but might be worth checking)

**Comment by Soare-Robert-Daniel** [comment]
>   How it looks on iphone: https://vertis.d.pr/WNU3dX
>   How it looks on android: https://vertis.d.pr/AqguPI
>   (I could replicate it only on Browserstack using iphone 11, i don't know if it's a real problem but might be worth checking)

Very weird error. I will look into it to see if it is a safari problem

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, take this ZIP for testing
[otter-blocks-dev.zip](https://github.com/Codeinwp/otter-blocks/files/9007718/otter-blocks-dev.zip)

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Found a single problem:

- [ ] I imported a form with all the settings via json file and when i tried to edit the post, the block was broken and had an error in console. If i disable cache and refresh the block works fine in editor, but i have to do it each time i click on **Edit post**. 
I made a video to better outline the issue https://vertis.d.pr/xSObFq

After this is fixed, it can be moved to **Ready to merge** 🚀

**Comment by Soare-Robert-Daniel** [comment]
Build to test.
[otter-blocks-dev.zip](https://github.com/Codeinwp/otter-blocks/files/9010428/otter-blocks-dev.zip)

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel The issue is fixed now, thank you 🚀

**Pull request merged**